[[ Bug 21396 ]] Fix crash on startup in iOS 12 beta

montegoulding opened the pull request:
This patch fixes an issue where the view layer appears to be retaining the
`CGDataProvider` from an image when it is drawn via `CGContextDrawImage`.
As we were freeing the raster directly after the draw the `CALayer` could not
access it from the `CGDataProvider` when rendering causing a crash.

This patch therefore changes the behavior of `MCGRasterCreateCGDataProvider`
where previously when calling it with the copy parameter `false` it would be
left to the caller to free the buffer the data provider now takes ownership
of it and it is freed via the data provider free callback. As a consequence
of the change custom cursor creation on mac now copies the buffer for
simplicity.

runrevmark commented on engine/src/mbliphonegfx.mm:
```diff
@@ -582,6 +581,8 @@ void FlushBits(MCGIntegerRectangle p_area, void *p_bits, uint32_t p_stride)
 		
 		glDeleteTextures(1, &t_texture);
+        
+        free(p_bits);
```
This seems a bit dodgy - surely the bits are owned by the MCGRaster which is passed in at a level above... Shouldn't the original caller be responsible for freeing? (assuming they aren't borrowed).

montegoulding replied:
Hmm... I had just moved the free from `UnlockPixels` to here seeing as `FlushBits` in `MCUIKitStackSurface` takes ownership of them so I wanted `FlushBits` to take them in both cases.

montegoulding replied:
Can you think of an alternative that doesn't involve copying the buffer in `FlushBits` for `MCUIKitStackSurface` given the `CGDataProvider` is expecting them to be available.

montegoulding commented on engine/src/mbliphonegfx.mm:
```diff
@@ -200,7 +200,6 @@ void UnlockPixels(MCGIntegerRectangle p_area, MCGRaster& p_raster, bool p_update
 		if (p_update)
 			FlushBits(p_area, p_raster . pixels, p_raster . stride);
 		
-		free(p_raster . pixels);
```
I see that I'm leaking at the moment here in the case of `! p_update`.. should now be:
```
if (p_update)
{
		FlushBits(p_area, p_raster . pixels, p_raster . stride);
}
else
{
		free(p_raster . pixels);
}
```

montegoulding replied:
@runrevmark perhaps `FlushBits` should set a boolean indicating if it has taken ownership of the bits? Then we can do:

```
bool t_taken = false;
if (p_update)
 		FlushBits(p_area, p_raster . pixels, p_raster . stride, t_taken);

if (!t_taken)
 		free(p_raster . pixels);
```

montegoulding replied:
@runrevmark I'm running with the above as it means I can only set x_taken if the raster is successfully converted to a CGImage. Will push up in a bit.

montegoulding committed:
[[ Bug 21396 ]] Fix crash on startup in iOS 12 beta

This patch fixes an issue where the view layer appears to be retaining the
`CGDataProvider` from an image when it is drawn via `CGContextDrawImage`.
As we were freeing the raster directly after the draw the `CALayer` could not
access it from the `CGDataProvider` when rendering causing a crash.

This patch therefore changes the behavior of `MCGRasterCreateCGDataProvider`
where previously when calling it with the copy parameter `false` it would be
left to the caller to free the buffer the data provider now takes ownership
of it and it is freed via the data provider free callback. As a consequence
of the change custom cursor creation on mac now copies the buffer for
simplicity.

runrevmark commented:
Okay so I think I have now stared at this too much! From what you said it fixes the issue and the code looks correct. So...

runrevmark commented:
@livecode-vulcan review ok 964d263

livecode-vulcan commented:
💙 review by @runrevmark ok 964d263d729b746dae105d777af2a492909f1505

livecode-vulcan commented:
😎 test success 964d263d729b746dae105d777af2a492909f1505

* try-community-armv6-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-sdk26_ndk16r15/builds/57)
* try-community-armv7-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-armv7-android-sdk26_ndk16r15/builds/49)
* try-community-arm64-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-arm64-android-sdk26_ndk16r15/builds/49)
* try-community-x86-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-x86-android-sdk26_ndk16r15/builds/49)
* try-community-x86_64-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-x86_64-android-sdk26_ndk16r15/builds/49)
* try-community-js-emscripten-sdk1.35: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/1571)
* try-community-universal-ios-iphoneos11.4: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos11.4/builds/98)
* try-community-universal-ios-iphonesimulator11.4: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator11.4/builds/98)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/1601)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/1595)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/1591)
* try-community-x86-win32: [success](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/1628)
* try-community-x86_64-win32: [success](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/1568)
